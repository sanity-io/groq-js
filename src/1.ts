export {evaluate} from './evaluator'
export type {GroqFunction, GroqFunctionArg, GroqPipeFunction} from './evaluator/functions'
export type {Scope} from './evaluator/scope'
export type {
  Context,
  DereferenceFunction,
  Document,
  EvaluateOptions,
  Executor,
} from './evaluator/types'
export * from './nodeTypes'
export {parse} from './parser'
export type {ParseOptions} from './types'
export type {
  AnyStaticValue,
  ArrayValue,
  BooleanValue,
  DateTimeValue,
  GroqType,
  NullValue,
  NumberValue,
  ObjectValue,
  Path,
  PathValue,
  StaticValue,
  StreamValue,
  StringValue,
  Value,
} from './values'
export {DateTime} from './values'
