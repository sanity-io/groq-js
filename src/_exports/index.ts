export {evaluateQuery as evaluate} from '../evaluator/evaluate'
export type {GroqFunction, GroqFunctionArg, GroqPipeFunction, ParseOptions} from '../types'
export * from '../nodeTypes'
export {parse} from '../parser/parser'
export {typeEvaluate} from '../typeEvaluator/typeEvaluate'
export {createReferenceTypeNode} from '../typeEvaluator/typeHelpers'

export type {
  ArrayTypeNode,
  BooleanTypeNode,
  Document as DocumentSchemaType,
  InlineTypeNode,
  NullTypeNode,
  NumberTypeNode,
  ObjectAttribute,
  ObjectTypeNode,
  PrimitiveTypeNode,
  Schema as SchemaType,
  StringTypeNode,
  TypeDeclaration as TypeDeclarationSchemaType,
  TypeNode,
  UnionTypeNode,
  UnknownTypeNode,
} from '../typeEvaluator/types'
