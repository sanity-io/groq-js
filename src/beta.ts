// Beta/experimental features - API subject to change

// Serializer exports
export {serialize} from './serializer'
export type {SerializeOptions} from './serializer'

// Node types (needed by ExprNode)
export * from './nodeTypes'

// Function types (needed by some nodes)
export type {GroqFunction, GroqFunctionArg, GroqPipeFunction} from './evaluator/functions'

// Evaluation types (needed by function types)
export type {Scope} from './evaluator/scope'
export type * from './evaluator/types'
export type * from './values'
