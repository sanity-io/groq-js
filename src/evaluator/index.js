const {
  StaticValue,
  StreamValue,
  MapperValue,
  NULL_VALUE,
  TRUE_VALUE,
  FALSE_VALUE,
  Range,
  Pair,
  fromNumber,
  fromJS
} = require('./value')
const {functions, pipeFunctions} = require('./functions')
const operators = require('./operators')

function inMapper(value, fn) {
  if (value instanceof MapperValue) {
    return new MapperValue(
      new StreamValue(async function*() {
        for await (let elementValue of value) {
          yield await fn(elementValue)
        }
      })
    )
  } else {
    return fn(value)
  }
}

class Scope {
  constructor(params, source, value, parent) {
    this.params = params
    this.source = source
    this.value = value
    this.parent = parent
    this.timestamp = parent ? parent.timestamp : new Date().toISOString()
  }

  createNested(value) {
    return new Scope(this.params, this.source, value, this)
  }
}

function execute(node, scope) {
  const func = EXECUTORS[node.type]
  if (!func) throw new Error('No executor for node.type=' + node.type)
  return func(node, scope)
}

const EXECUTORS = {
  This(_, scope) {
    return scope.value
  },

  Star(_, scope) {
    return scope.source
  },

  Parameter({name}, scope) {
    return fromJS(scope.params[name])
  },

  Parent({n}, scope) {
    for (let i = 0; i < n; i++) {
      scope = scope.parent
      if (!scope) return NULL_VALUE
    }
    return scope.value
  },

  OpCall({op, left, right}, scope) {
    let func = operators[op]
    if (!func) throw new Error('Unknown operator: ' + op)
    return func(left, right, scope, execute)
  },

  FuncCall({name, args}, scope) {
    let func = functions[name]
    if (!func) throw new Error('Unknown function: ' + name)
    return func(args, scope, execute)
  },

  async PipeFuncCall({base, name, args}, scope) {
    let func = pipeFunctions[name]
    if (!func) throw new Error('Unknown function: ' + name)
    let baseValue = await execute(base, scope)
    return func(baseValue, args, scope, execute)
  },

  async Filter({base, query}, scope) {
    let baseValue = await execute(base, scope)

    return inMapper(baseValue, async value => {
      if (value.getType() != 'array') return NULL_VALUE

      return new StreamValue(async function*() {
        for await (let element of value) {
          let newScope = scope.createNested(element)
          let condValue = await execute(query, newScope)
          if (condValue.getBoolean()) yield element
        }
      })
    })
  },

  async Element({base, index}, scope) {
    let baseValue = await execute(base, scope)

    return inMapper(baseValue, async arrayValue => {
      if (arrayValue.getType() != 'array') return NULL_VALUE

      let idxValue = await execute(index, scope)
      if (idxValue.getType() != 'number') return NULL_VALUE

      // OPT: Here we can optimize when idx >= 0
      let array = await arrayValue.get()
      let idx = await idxValue.get()

      if (idx < 0) {
        idx = array.length + idx
      }

      if (idx >= 0 && idx < array.length) {
        return new StaticValue(array[idx])
      } else {
        // Make sure we return `null` for out-of-bounds access
        return NULL_VALUE
      }
    })
  },

  async Slice({base, left, right, isExclusive}, scope) {
    let baseValue = await execute(base, scope)

    return inMapper(baseValue, async arrayValue => {
      if (arrayValue.getType() != 'array') return NULL_VALUE

      let leftIdxValue = await execute(left, scope)
      let rightIdxValue = await execute(right, scope)

      if (leftIdxValue.getType() != 'number' || rightIdxValue.getType() != 'number') {
        return NULL_VALUE
      }

      // OPT: Here we can optimize when either indices are >= 0
      let array = await arrayValue.get()
      let leftIdx = await leftIdxValue.get()
      let rightIdx = await rightIdxValue.get()

      // Handle negative index
      if (leftIdx < 0) leftIdx = array.length + leftIdx
      if (rightIdx < 0) rightIdx = array.length + rightIdx

      // Convert from inclusive to exclusive index
      if (!isExclusive) rightIdx++

      if (leftIdx < 0) leftIdx = 0
      if (rightIdx < 0) rightIdx = 0

      // Note: At this point the indices might point out-of-bound, but
      // .slice handles this correctly.

      return new StaticValue(array.slice(leftIdx, rightIdx))
    })
  },

  async Attribute({base, name}, scope) {
    let baseValue = await execute(base, scope)

    return inMapper(baseValue, async value => {
      if (value.getType() == 'object') {
        let data = await value.get()
        if (data.hasOwnProperty(name)) {
          return new StaticValue(data[name])
        }
      }

      return NULL_VALUE
    })
  },

  async Identifier({name}, scope) {
    if (scope.value.getType() == 'object') {
      let data = await scope.value.get()
      if (data.hasOwnProperty(name)) {
        return new StaticValue(data[name])
      }
    }

    return NULL_VALUE
  },

  Value({value}) {
    return new StaticValue(value)
  },

  async Mapper({base}, scope) {
    let baseValue = await execute(base, scope)
    if (baseValue.getType() != 'array') return baseValue

    if (baseValue instanceof MapperValue) {
      return new MapperValue(
        new StreamValue(async function*() {
          for await (let element of baseValue) {
            if (element.getType() == 'array') {
              for await (let subelement of element) {
                yield subelement
              }
            } else {
              yield NULL_VALUE
            }
          }
        })
      )
    } else {
      return new MapperValue(baseValue)
    }
  },

  async Parenthesis({base}, scope) {
    let baseValue = await execute(base, scope)
    if (baseValue instanceof MapperValue) {
      baseValue = baseValue.value
    }
    return baseValue
  },

  async Projection({base, query}, scope) {
    let baseValue = await execute(base, scope)
    return inMapper(baseValue, async baseValue => {
      if (baseValue.getType() == 'null') return NULL_VALUE

      if (baseValue.getType() == 'array') {
        return new StreamValue(async function*() {
          for await (let value of baseValue) {
            let newScope = scope.createNested(value)
            let newValue = await execute(query, newScope)
            yield newValue
          }
        })
      } else {
        let newScope = scope.createNested(baseValue)
        return await execute(query, newScope)
      }
    })
  },

  async Deref({base}, scope) {
    let baseValue = await execute(base, scope)
    return inMapper(baseValue, async baseValue => {
      if (scope.source.getType() != 'array') return NULL_VALUE
      if (baseValue.getType() != 'object') return NULL_VALUE

      let id = (await baseValue.get())._ref
      if (typeof id != 'string') return NULL_VALUE

      for await (let doc of scope.source) {
        if (id === doc.data._id) {
          return doc
        }
      }

      return NULL_VALUE
    })
  },

  async Object({attributes}, scope) {
    let result = {}
    for (let attr of attributes) {
      switch (attr.type) {
        case 'ObjectAttribute': {
          let key = await execute(attr.key, scope)
          if (key.getType() != 'string') continue

          let value = await execute(attr.value, scope)
          if (value.getType() == 'null') {
            delete result[key.data]
            break
          }

          result[key.data] = await value.get()
          break
        }

        case 'ObjectConditionalSplat': {
          let cond = await execute(attr.condition, scope)
          if (!cond.getBoolean()) continue

          let value = await execute(attr.value, scope)
          if (value.getType() != 'object') continue
          Object.assign(result, value.data)
          break
        }

        case 'ObjectSplat': {
          let value = await execute(attr.value, scope)
          if (value.getType('object')) {
            Object.assign(result, value.data)
          }
          break
        }

        default:
          throw new Error('Unknown node type: ' + attr.type)
      }
    }
    return new StaticValue(result)
  },

  Array({elements}, scope) {
    return new StreamValue(async function*() {
      for (let element of elements) {
        let value = await execute(element.value, scope)
        if (element.isSplat) {
          if (value.getType() == 'array') {
            for await (let v of value) {
              yield v
            }
          }
        } else {
          yield value
        }
      }
    })
  },

  async Range({left, right, isExclusive}, scope) {
    let leftValue = await execute(left, scope)
    let rightValue = await execute(right, scope)

    if (!Range.isConstructible(leftValue.getType(), rightValue.getType())) {
      return NULL_VALUE
    }

    let range = new Range(await leftValue.get(), await rightValue.get(), isExclusive)
    return new StaticValue(range)
  },

  async Pair({left, right}, scope) {
    let leftValue = await execute(left, scope)
    let rightValue = await execute(right, scope)

    let pair = new Pair(await leftValue.get(), await rightValue.get())
    return new StaticValue(pair)
  },

  async Or({left, right}, scope) {
    let leftValue = await execute(left, scope)
    let rightValue = await execute(right, scope)

    if (leftValue.getType() == 'boolean') {
      if (leftValue.data == true) return TRUE_VALUE
    }

    if (rightValue.getType() == 'boolean') {
      if (rightValue.data == true) return TRUE_VALUE
    }

    if (leftValue.getType() != 'boolean') return NULL_VALUE
    if (rightValue.getType() != 'boolean') return NULL_VALUE

    return FALSE_VALUE
  },

  async And({left, right}, scope) {
    let leftValue = await execute(left, scope)
    let rightValue = await execute(right, scope)

    if (leftValue.getType() == 'boolean') {
      if (leftValue.data == false) return FALSE_VALUE
    }

    if (rightValue.getType() == 'boolean') {
      if (rightValue.data == false) return FALSE_VALUE
    }

    if (leftValue.getType() != 'boolean') return NULL_VALUE
    if (rightValue.getType() != 'boolean') return NULL_VALUE

    return TRUE_VALUE
  },

  async Not({base}, scope) {
    let value = await execute(base, scope)
    if (value.getType() != 'boolean') {
      return NULL_VALUE
    }
    return value.getBoolean() ? FALSE_VALUE : TRUE_VALUE
  },

  async Neg({base}, scope) {
    let value = await execute(base, scope)
    if (value.getType() != 'number') return NULL_VALUE
    return fromNumber(-(await value.get()))
  },

  async Pos({base}, scope) {
    let value = await execute(base, scope)
    if (value.getType() != 'number') return NULL_VALUE
    return fromNumber(await value.get())
  },

  async Asc() {
    return NULL_VALUE
  },

  async Desc() {
    return NULL_VALUE
  }
}

/**
 * Evaluates a syntax tree (which you can get from {@link module:groq-js.parse}).
 *
 * @param {SyntaxNode} tree
 * @param {object} [options] Options.
 * @param {object} [options.params]  Parameters availble in the GROQ query (using `$param` syntax).
 * @param [options.root] The value that will be available as `@` in GROQ.
 * @param [options.dataset] The value that will be available as `*` in GROQ.
 * @return {Value}
 * @alias module:groq-js.evaluate
 */
async function evaluate(tree, options = {}) {
  let root = fromJS(options.root)
  let dataset = fromJS(options.dataset)
  let params = {}

  if (options.params) {
    Object.assign(params, options.params)
  }

  let scope = new Scope(params, dataset, root, null)
  return await execute(tree, scope)
}

exports.evaluate = evaluate
