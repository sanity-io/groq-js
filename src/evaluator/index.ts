export {tryConstantEvaluate} from './constantEvaluate'
export {evaluateQuery as evaluate} from './evaluate'
