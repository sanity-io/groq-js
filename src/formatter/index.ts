import type {ExprNode} from '../nodeTypes'
import {createFormatContext} from './context'
import {NodeFormatter} from './nodeFormatters'

export interface FormatOptions {
  /** Indentation string (default: "  ") */
  indent?: string
}

/**
 * Format a GROQ AST node into pretty format
 */
export function format(node: ExprNode, options: FormatOptions = {}): string {
  const context = createFormatContext(options)
  const formatter = new NodeFormatter(context)
  return formatter.format(node)
}
