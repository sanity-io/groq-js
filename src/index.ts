export {parse} from './parser'
export {evaluate} from './evaluator'
