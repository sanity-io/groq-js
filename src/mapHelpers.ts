import {CompoundMapper, Mapper, SyntaxNode} from './nodeTypes'

export type MapState = {
  base: SyntaxNode
  mode: 'array' | 'simple'
  mapper: CompoundMapper
}

export function switchMapState(
  base: SyntaxNode,
  state: MapState | undefined,
  mode: 'array' | 'simple'
): MapState {
  if (!state) {
    let mapper: CompoundMapper = {type: 'Compound', mappers: []}

    return {
      base: {
        type: 'Map',
        base,
        mapper
      },
      mode,
      mapper
    }
  }

  if (state.base !== base) throw new Error('switchMapState got incorrect base')

  if (state.mode === mode) return state

  let mapper: CompoundMapper = {type: 'Compound', mappers: []}
  if (mode === 'array') {
    state.mapper.mappers.push(mapper)
  } else {
    state.mapper.mappers.push({type: 'Apply', mapper: mapper})
  }

  state.mode = mode
  state.mapper = mapper
  return state
}

export function addMapper(state: MapState, mapper: Mapper) {
  state.mapper.mappers.push(mapper)
}
