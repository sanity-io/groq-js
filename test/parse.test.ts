import {parse} from '../src/1'

import t from 'tap'
import {throwsWithMessage} from './testUtils'

t.test('Basic parsing', async (t) => {
  t.test('Example query', async (t) => {
    const query = `*[_type == "product"]{name}`
    const tree = parse(query)
    t.matchSnapshot(tree)
  })

  t.test('Trailing comma in function call', async (t) => {
    const query = `select(123,)`
    const tree = parse(query)
    t.matchSnapshot(tree)
  })

  t.test('Object expression starting with string', async (t) => {
    const query = `{"mail" == 1 => {}}`
    const tree = parse(query)
    t.matchSnapshot(tree)
  })

  t.test('Space after field in objects', async (t) => {
    const query = `{"mail" : 123}`
    const tree = parse(query)
    t.matchSnapshot(tree)
  })

  t.test('Comment with no text', async (t) => {
    const query = `
      // my comment
      //
      1
    `
    const tree = parse(query)
    t.matchSnapshot(tree)
  })

  t.test('Complex query', async (t) => {
    const query = `count(array::unique(
      *[_type == 'page']{"_id": select(
        _id in path("drafts.**") => _id,
        "drafts." + _id
      )}._id
    ))`

    const tree = parse(query)
    t.matchSnapshot(tree)
  })
})

t.test('Error reporting', async (t) => {
  t.test('when querying with a syntax error', async (t) => {
    t.plan(3)
    const query = `*[_type == "]`
    try {
      parse(query)
    } catch (error: any) {
      t.same(error.name, 'GroqSyntaxError')
      t.same(error.position, 13)
      t.same(error.message, 'Syntax error in GROQ query at position 13')
    }
  })
})

t.test('Expression parsing', async (t) => {
  t.test('when parsing strings', async (t) => {
    t.test('handles unicode hex characters', async (t) => {
      const queries = ['*[foo == "\\u{0040}"]', '*[foo == "\\u0040"]']

      for (const query of queries) {
        t.doesNotThrow(() => parse(query))
      }
    })
  })

  t.test('when parsing objects', async (t) => {
    t.test('throws when the object is not terminated properly', async (t) => {
      throwsWithMessage(t, () => parse('*{a, b'), 'Syntax error in GROQ query at position 5')
    })
  })

  t.test('when parsing functions', async (t) => {
    t.test('throws when the function call is not terminated properly', async (t) => {
      throwsWithMessage(t, () => parse('count(*[]'), 'Syntax error in GROQ query at position 9')
    })

    t.test('throws when using boost() when `allowBoost` is false', async (t) => {
      throwsWithMessage(t, () => parse('boost()'), 'unexpected boost')
    })

    t.test('throws when an undefined namespace is used', async (t) => {
      throwsWithMessage(t, () => parse('invalid::func()'), 'Undefined namespace: invalid')
    })
  })

  t.test('throws when nothing is passed', async (t) => {
    throwsWithMessage(t, () => parse(''), 'Syntax error in GROQ query at position 0')
  })

  t.test('has support for tuples', async (t) => {
    t.doesNotThrow(() => parse('(a, b, c)'))
  })

  t.test('throws when an open square bracket has no closing match', async (t) => {
    throwsWithMessage(t, () => parse('*[foo == [1, 2'), 'Syntax error in GROQ query at position 13')
  })

  t.test('when parsing pipecalls', async (t) => {
    t.test('throws when using a namespace other than `global`', async (t) => {
      throwsWithMessage(t, () => parse('* | invalid::func()'), 'Undefined namespace: invalid')
    })

    t.test('throws when using an invalid function', async (t) => {
      throwsWithMessage(t, () => parse('* | func()'), 'Undefined pipe function: func')
    })
  })

  t.test('when parsing `desc`', async (t) => {
    t.test('throws when used unexpectedly', async (t) => {
      throwsWithMessage(t, () => parse('*[_type desc]'), 'unexpected desc')
    })
  })

  t.test('when parsing slices', async (t) => {
    t.test('throws when a constant number is not used', async (t) => {
      throwsWithMessage(t, () => parse('*[0..x]'), 'slicing must use constant numbers')
    })
  })

  t.test('when extracting property keys', async (t) => {
    t.test('throws when the key cannot be determined', async (t) => {
      throwsWithMessage(t, () => parse('*{1}'), 'Cannot determine property key for type: Value')
    })
  })
})

t.test('Selector validation', async (t) => {
  t.test('passes with valid selectors', async (t) => {
    const queries = [
      'diff::changedAny({}, {}, foo)',
      'diff::changedAny({}, {}, foo.bar.baz)',
      'diff::changedAny({}, {}, foo[])',
      'diff::changedAny({}, {}, foo.bar[baz == 1])',
      'diff::changedAny({}, {}, (foo))',
      'diff::changedAny({}, {}, anywhere(foo))',
    ]

    for (const query of queries) {
      t.doesNotThrow(() => parse(query))
    }
  })

  t.test('fails with invalid selectors', async (t) => {
    const queries = [
      'diff::changedAny({}, {}, "foo")',
      'diff::changedAny({}, {}, 1 + 2)',
      'diff::changedAny({}, {}, 5)',
      'diff::changedAny({}, {}, (foo, bar))', // BUG: fails because we don't support tuples inside traversals yet
    ]

    for (const query of queries) {
      throwsWithMessage(t, () => parse(query), 'Invalid selector syntax')
    }
  })

  // BUG: the below cases throw syntax errors because we don't support their syntax *yet*.
  // These are valid selectors according to the spec, but not according to our implementation.
  t.test('fails due to syntax error', async (t) => {
    const queries = [
      'diff::changedAny({}, {}, a.(foo).bar)',
      'diff::changedAny({}, {}, a[].(foo, bar).b)',
    ]

    for (const query of queries) {
      try {
        parse(query)
      } catch (error: any) {
        t.match(error.message, 'Syntax error in GROQ query at position')
      }
    }
  })
})

t.test('Delta-GROQ', async (t) => {
  const queries = [
    `before().title == after().title`,
    `delta::changedAny(name)`,
    `delta::changedAny((name, description))`,
    `delta::changedAny((name, description)._type)`,
    `delta::changedOnly(foo)`,
    `delta::changedOnly(foo[bar == 1])`,
  ]

  for (const query of queries) {
    t.test(query, async (t) => {
      t.throws(() => parse(query))
      t.doesNotThrow(() => parse(query, {mode: 'delta'}))
    })
  }
})
