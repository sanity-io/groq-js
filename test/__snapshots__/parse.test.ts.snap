// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic parsing Example query 1`] = `
Object {
  "base": Object {
    "base": Object {
      "type": "Everything",
    },
    "expr": Object {
      "left": Object {
        "name": "_type",
        "type": "AccessAttribute",
      },
      "op": "==",
      "right": Object {
        "type": "Value",
        "value": "product",
      },
      "type": "OpCall",
    },
    "type": "Filter",
  },
  "expr": Object {
    "base": Object {
      "type": "This",
    },
    "expr": Object {
      "attributes": Array [
        Object {
          "name": "name",
          "type": "ObjectAttributeValue",
          "value": Object {
            "name": "name",
            "type": "AccessAttribute",
          },
        },
      ],
      "type": "Object",
    },
    "type": "Projection",
  },
  "type": "Map",
}
`;
